README.md
# timetable-checker
Timetable checker for CityU mates. Reads class information from AIMS and match against your timetable to see if sections conflit occurs.


changelog.md
# CityU Timetable Checker

&copy; 2020-2021 Kelvin Lee. All rights reserved.

## Version 0.3.1 (Released on 27/08/2021)

**New functions:**

- Add a context menu (right-click menu) item that can jump to Master Class Schedule immediately.
![Example image of context menu item](context.png)
- Adjusted auto-filling function. Sections which are full will not be auto-filled.

**Fixes:**

- No response when adding CRNs manually to the wishlist.

## Version 0.3.0 (Released on 30/12/2020)

**New functions:**

- As face-to-face teaching mode will be used more widely in Semester B 2020/21, the timetable view will now show the location where the lesson is conducted. Also, it will show the name of instructor of that section.
- Add warning when non-web-enabled sections exists in the wishlist, and exclude non-web-enabled sections from auto-fill function.

**Fixes:**

- Error occured when there are wishlist items which does not have status info (CRNs added manually in the popup window).

## Version 0.2.0 (Released on 31/08/2020)

**New functions:**

- Updated data structure of wishlist so that more information can be shown:
    - Course code and Section code*
    - Available seats / Waitlist remaining quota*
    - Conflicts*
    - Updated time
- Timetable view (on the popup and Timetable preview) are adjusted.
- Action cells split to 2, showing the two actions seperately.
- Changed page action to browser action so timetable can be accessed even not in AIMS. It also adds the ability to show the number of CRNs added to wishlist.

\* If the extension found a CRN matching with the wishlist when you browse the sections list of the Master Class Schedule, the information will be updated automatically.

**Fixes:**

- "Course registered" status may not be correctly shown if the section has more than one timeslots.

## Version 0.1.0 (Released on 29/08/2020)

This is the first release version of this extension.

js/bg.js
/* chrome.tabs.onUpdated.addListener(function (i, c, t) {
    chrome.tabs.query({ url: [ "https://banweb.cityu.edu.hk/*" ] }, function(tabs) {
        for (i = 0; i < tabs.length; i++) { 
            chrome.pageAction.show(tabs[i].id);
        }
    });
}); */

chrome.contextMenus.create({
    "title": "Go to Master Class Schedule",
    "id": "nav-mcs",
    "contexts": ["page"],
    "onclick": function (e, t) {
        chrome.tabs.sendMessage(t.id, { action: "mcs" });
    },
    "documentUrlPatterns": ["https://banweb.cityu.edu.hk/*"]
}, function () {
    if (chrome.runtime.lastError) console.error(chrome.runtime.lastError.message);
});

chrome.runtime.onMessage.addListener(({type, data}, s) => {
    if (type === "preview") {
        chrome.storage.local.set({ preview: data }, () => {
            chrome.windows.create({
                url: "preview.html",
                type: "popup",
                focused: true
            });
        });
    }
});

chrome.storage.local.get("wishlist", ({ wishlist }) => {
    if (wishlist) {
        chrome.browserAction.setBadgeText({
            text: `${wishlist.length}`
        });
    }
    chrome.browserAction.setBadgeBackgroundColor({
        color: "#68768a"
    });
})

chrome.storage.onChanged.addListener(({ wishlist }, n) => {
    if (n === "local" && wishlist != null) {
        chrome.browserAction.setBadgeText({
            text: `${wishlist.newValue.length}`
        });
    }
});

js/popup.js
const tabBar = new mdc.tabBar.MDCTabBar(document.querySelector(".mdc-tab-bar")),
    t_addCrn = new mdc.textField.MDCTextField(document.querySelector("#add-crn")),
    list = new mdc.list.MDCList(document.querySelector("#crn-list")),
    snackbar = new mdc.snackbar.MDCSnackbar(document.querySelector('.mdc-snackbar')),
    s_autofill = new mdc.switchControl.MDCSwitch(document.querySelector('#autofill-switch-container')),
    d_details = new mdc.dialog.MDCDialog(document.querySelector("#section-details-dialog")),
    START_DATE = "2021-07-27 08:30:00",
    END_DATE = "2021-09-06 23:30:00",
    DAYS = ["", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];

snackbar.timeoutMs = 4000;

if (!(moment().isBetween(START_DATE, END_DATE, undefined, "[]"))) {
    $(".mdc-tab-bar").hide();
}

document.querySelectorAll(".mdc-button").forEach((v, i) => {
    mdc.ripple.MDCRipple.attachTo(v);
});

for (let i = 0; i < 16; i++) {
    for (let j = 0; j < 7; j++) {
        let div$ = $("<div></div");
        if (j >= 5) {
            div$.addClass("weekend");
        }
        $(".timetable .background").append(div$);
    }
}

tabBar.listen("MDCTabBar:activated", ({ detail: { index } }) => {
    $(".tabbed-content").hide();
    $(`.tabbed-content:eq(${index})`).show();
});

s_autofill.listen("change", () => {
    chrome.storage.local.set({ autofill: s_autofill.checked }, () => {
        snackbar.labelText = "Auto-fill setting changed.";
        snackbar.open();
    });
});

chrome.storage.local.get(["ttb", "wishlist", "autofill"], ({ ttb, wishlist, autofill }) => {
    s_autofill.checked = autofill || false;

    if (ttb != null) {
        let m1 = moment(ttb.meta[2], "MMM DD, YYYY h:mm a");
        $("#meta-text").text(`${ttb.meta[0]}, ${ttb.meta[1]}, updated at ${m1.format("dddd, Do MMMM, YYYY HH:mm")}`);
        ttb.classes.forEach((i) => {
            i.times.forEach(j => {
                if (j === null || j.time === null) {
                    let div$ = $("<div></div>");
                    div$.html(`${i.name[1].replace(" ", "")} ${i.name[2]}`);
                    div$.attr("title", `${i.name[0]}\n${i.crn}\n${i.instruct}`);
                    div$.attr("data-crse", `${i.name[1]},${i.name[2]}`);
                    div$.click(showDetails);
                    $("#no-scheduled-time").show().append(div$);
                    return;
                }
                let times = j.time.map(t => new moment(t, "hh:mm a"));
                let id = `ttb-${j.day}-${times[0].hour() - 7}-${times[1].hour() - 6}`;
                if ($(`#${id}`).length === 0) {
                    let div$ = $("<div></div>");
                    div$.html(`${i.name[1].replace(" ", "")} <small>${i.name[2]}</small><br /><small>${j.loc}</small>`);
                    div$.attr("title", `${i.name[0]}\n${i.crn}\n${times[0].format("HH:mm")} - ${times[1].format("HH:mm")}\n${i.instruct}`);
                    div$.attr("id", id);
                    div$.attr("data-crse", `${i.name[1]},${i.name[2]}`);
                    div$.css({
                        "grid-row-start": `${times[0].hour() - 7}`,
                        "grid-row-end": `${times[1].hour() - 6}`,
                        "grid-column-start": `${j.day}`
                    });
                    div$.click(showDetails);
                    $(".timetable .content").append(div$);
                }
            });
        });
    }

    if (wishlist != null) {
        if (!wishlist.every(({ status }) => {
            return status == null || status.webenabled;
        })) {
            $("#has-non-web-enabled").show();
        }
        wishlist.forEach(({ crn, status }) => {
            let n = document.importNode(document.querySelector("#crn-item-template").content, true);
            $(n.children[0]).find(".mdc-checkbox__native-control").attr("id", `crn-${crn}`);
            $(n.children[0]).find(".mdc-list-item__text").attr("for", `crn-${crn}`);
            if (status == null) {
                $(n.children[0]).find(".mdc-list-item__primary-text").text(`${crn}`);
                $(n.children[0]).find(".mdc-list-item__secondary-text").append($("<span></span>").css("color", "red").text("Check Master Class Schedule for details."));
            } else {
                const { course, section, webenabled, avail, cap, waitlist, conflict, updated } = status;
                $(n.children[0]).find(".mdc-list-item__primary-text").text(`${crn} (${course} ${section})`);
                if (webenabled) {
                    let warn = false;
                    if (avail === 0) {
                        if (waitlist === false) {
                            warn = true;
                            $(n.children[0]).find(".mdc-list-item__secondary-text").append($("<span></span>").css("color", "red").text(`Section is full, waitlist not available`));
                        } else if (waitlist === 0) {
                            warn = true;
                            $(n.children[0]).find(".mdc-list-item__secondary-text").append($("<span></span>").css("color", "red").text(`Section is full, waitlist full`));
                        } else {
                            $(n.children[0]).find(".mdc-list-item__secondary-text").append($("<span></span>").css("color", "darkorange").text(`Section is full, waitlist remaining ${waitlist}`));
                        }
                    } else {
                        $(n.children[0]).find(".mdc-list-item__secondary-text").append($("<span></span>").css("color", "green").text(`Available: ${avail} / ${cap}`));
                    }
                    $(n.children[0]).find(".mdc-list-item__secondary-text").append("; ");
                    if (conflict) {
                        warn = true;
                        $(n.children[0]).find(".mdc-list-item__secondary-text").append($("<span></span>").css("color", "red").text(`Conflicts with ${conflict}`));
                    } else {
                        $(n.children[0]).find(".mdc-list-item__secondary-text").append($("<span></span>").css("color", "green").text(`No conflicts`));
                    }
                    if (warn) {
                        $(n.children[0]).append(`<i class="material-icons mdc-list-item__meta" aria-hidden="true">warning</i>`);
                    }
                } else {
                    $(n.children[0]).append(`<i class="material-icons mdc-list-item__meta" aria-hidden="true">error</i>`);
                    $(n.children[0]).find(".mdc-list-item__secondary-text").append($("<span></span>").css("color", "red").text(`Section not web-enabled`));
                }
                $(n.children[0]).find(".mdc-list-item__secondary-text").append("; ");
                $(n.children[0]).find(".mdc-list-item__secondary-text").append(`Updated ${moment(updated).fromNow()}`);
            }
            $(n.children[0]).attr("data-crn", crn).appendTo("#crn-list");
        });

        document.querySelectorAll(".mdc-list-item").forEach((v, i) => {
            mdc.ripple.MDCRipple.attachTo(v);
        });

        document.querySelectorAll(".mdc-checkbox").forEach((v, i) => {
            mdc.checkbox.MDCCheckbox.attachTo(v);
        });

        list.layout();
    }
});

$("#crn-list").on("change", ".mdc-checkbox__native-control", (e) => {
    if (e.target.checked)
        $(e.target).parents("li").addClass("mdc-list-item--selected");
    else
        $(e.target).parents("li").removeClass("mdc-list-item--selected");
})

$("#empty-button").click(() => emptyWishlist());

$("#delete-button").click(() => deleteCRN());

$("#add-crn-button").click(() => {
    if (t_addCrn.valid && t_addCrn.value !== "") {
        addCRNToWishlist(parseInt(t_addCrn.value));
        t_addCrn.value = "";
    }
});

$("#add-crn-input").keydown((e) => {
    if (e.keyCode === 13) {
        if (t_addCrn.valid && t_addCrn.value !== "") {
            addCRNToWishlist(parseInt(t_addCrn.value));
            t_addCrn.value = "";
        }
    }
})

function showDetails(e) {
    chrome.storage.local.get("ttb", ({ ttb }) => {
        $("#section-details-table .mdc-data-table__content").empty();
        let le = ttb.classes.find(v => {
            let g = $(e.target).data("crse").split(",");
            return v.name[1] === g[0] && v.name[2] === g[1];
        });
        if (le == null) return;
        let details = [
            ["Course Code", le.name[1].replace(" ", "")],
            ["Course Name", le.name[0]],
            ["Section", le.name[2]],
            ["CRN", le.crn],
            ["Time", le.times.map(w => w?.time?.join(" - ") ?? "").join("<br />")],
            ["Days", le.times.map(w => w ? DAYS[w.day] : "").join("<br />")],
            ["Where", le.times.map(w => w?.loc ?? "").join("<br />")],
            ["Date Range", le.times.map(w => w?.range?.join(" - ") ?? "").join("<br />")],
            ["Instructor", le.instruct]
        ];
        $("#section-details-table .mdc-data-table__content").append(details.map(de => {
            let n = document.importNode(document.querySelector("#data-table-row-template").content, true);
            $(n.children[0]).find(".details-name").text(de[0]);
            $(n.children[0]).find(".details-value").html(de[1]);
            return $(n.children[0]);
        }))
        d_details.open();
    })
}

function deleteCRN() {
    chrome.storage.local.get("wishlist", ({ wishlist }) => {
        let sel$ = $("#crn-list .mdc-list-item.mdc-list-item--selected");
        if (sel$.length > 0) {
            sel$.each((i, li) => wishlist.splice(wishlist.findIndex(({ crn }) => crn === $(li).data("crn")), 1));
            chrome.storage.local.set({ wishlist: wishlist }, () => {
                snackbar.labelText = "Selected CRNs are removed from wishlist.";
                snackbar.open();
                sel$.remove();
                list.layout();
            });
        } else {
            snackbar.labelText = "No CRN selected.";
            snackbar.open();
        }
    });
}

function emptyWishlist() {
    chrome.storage.local.set({ wishlist: [] }, () => {
        snackbar.labelText = "Wishlist cleared.";
        snackbar.open();
        $("#crn-list").empty();
        list.layout();
    });
}

function addCRNToWishlist(crn) {
    chrome.storage.local.get(["ttb", "wishlist"], ({ ttb, wishlist }) => {
        if (wishlist == null) {
            wishlist = [];
        }
        if (ttb.classes.findIndex(v => v.crn === crn) !== -1) {
            snackbar.labelText = "This course is already registered.";
            snackbar.open();
            return;
        }
        if (wishlist.includes(crn)) {
            snackbar.labelText = "CRN already exists.";
            snackbar.open();
            return;
        }
        wishlist.push({ crn: crn, status: null });
        chrome.storage.local.set({ wishlist: wishlist }, () => {
            snackbar.labelText = "CRN added to wishlist.";
            snackbar.open();
            let n = document.importNode(document.querySelector("#crn-item-template").content, true);
            $(n.children[0]).find(".mdc-checkbox__native-control").attr("id", `crn-${crn}`);
            $(n.children[0]).find(".mdc-list-item__text").attr("for", `crn-${crn}`);
            $(n.children[0]).find(".mdc-list-item__primary-text").text(crn);
            $(n.children[0]).find(".mdc-list-item__secondary-text").text("No details saved; Please check Master Class Schedule for details.");
            $(n.children[0]).attr("data-crn", crn).appendTo("#crn-list");

            list.layout();

            document.querySelectorAll(".mdc-list-item").forEach((v, i) => {
                mdc.ripple.MDCRipple.attachTo(v);
            });

            document.querySelectorAll(".mdc-checkbox").forEach((v, i) => {
                mdc.checkbox.MDCCheckbox.attachTo(v);
            });
        });
    });
    return false;
}

js/read.js
const DAYS = " MTWRFS", FULL_REGEX = /.*f.*u.*l.*l.*/i, NUM_FILTER = /[^\d]/g,
    START_DATE = "2021-07-27 08:30:00",
    END_DATE = "2021-09-06 23:30:00",
    LOC_MAP = {
        "Lau Ming Wai Academic Building": "LAU",
        "Li Dak Sum Yip Yio Chin A Bldg": "LI",
        "Yeung Kin Man Acad Building": "YEUNG",
        "Mong Man Wai Building": "MMW"
    };

let inAddDropPeriod = moment().isBetween(START_DATE, END_DATE, undefined, "[]");

if (location.href === "https://banweb.cityu.edu.hk/pls/PROD/bwskfshd.P_CrseSchdDetl") {
    let classes = [];
    let meta = [];
    let z = $(".datadisplaytable");
    meta = $(".staticheaders").text().trim().split("\n");
    for (let i = 0; i < z.length; i += 2) {
        let j = {};
        j['name'] = z.eq(i).find("caption").text().split(" - ");
        j['crn'] = parseInt(z.eq(i).find("tr").eq(1).children().eq(1).text());
        j['instruct'] = z.eq(i).find("tr").eq(3).children().eq(1).text().trim();
        if (z.eq(i + 1).attr("summary") !== "This table lists the scheduled meeting times and assigned instructors for this class..") {
            j["times"] = [null];
            i--;
        } else {
            j['times'] = z.eq(i + 1).find("tr").slice(1).get().map(v => {
                return {
                    time: $(v).children().eq(1).text() === "TBA" ? null : $(v).children().eq(1).text().split(" - "),
                    day: $(v).children().eq(2).text().trim() === "" ? null : DAYS.indexOf($(v).children().eq(2).text()),
                    range: $(v).children().eq(4).text().split(" - "),
                    loc: $(v).children().eq(1).text() === "TBA" ? null : $(v).children().eq(3).text().trim().replace(/^(?<bldg>[\w ]*) (?<rm>LT-\d{1,2}|[A-Z]?\d-?\d{3})$/, function (v, bldg, rm) {
                        return `${LOC_MAP[bldg]} ${rm}`;
                    })
                };
            });
        }
        classes.push(j);
    }
    chrome.storage.local.set({ ttb: { meta, classes } }, () => {
        console.log("Saved timetable data", { meta, classes });
    });
} else if (location.href.startsWith("https://banweb.cityu.edu.hk/pls/PROD/hwscrssh_cityu.P_DispOneSection")) {
    chrome.storage.local.get(["ttb", "wishlist"], ({ ttb, wishlist }) => {
        let ctd$;
        let params = new URLSearchParams(location.search);
        let selected = /Courses Offered in  (Semester [AB] \d{4}\/\d{2}|Summer \d{4})/.exec($(".pagetitlediv h2").text())[1];
        let current = ttb.meta[1];
        if (selected !== current) {
            let f$ = $("<div></div>").css({ "font-size": "18px", "padding": "8px", "position": "fixed", "top": "20px", "right": "20px", "left": "20px", "text-align": "center", "background-color": "white", "border": "5px solid orange" })
                .text(`Term selected in Master Class Schedule is not compatible with your saved schedule (Required: ${selected}; Found: ${current})`)
                .appendTo(document.body);
        }
        $(".body table[border] tr").each((i, v) => {
            if (i === 0) {
                let th1$ = $("<th></th>").text("Status");
                let th2$ = $("<th></th>").text("Actions");
                if (inAddDropPeriod)
                    th2$.attr("colspan", "2");
                $(v).append([th1$, th2$]);
            } else {
                let flag = true;
                let crash = null;
                let wlitem = null;
                let wlidx = 0;
                let td1$ = $("<td></td>");
                let td2$ = $("<td></td>");
                let td3$ = $("<td></td>");
                if ($(v).is("[bgcolor='#ffccff']")) {
                    if ($(v).children().eq(6).text().trim() !== "") {
                        ctd$ = $(v);
                        wlitem = wishlist.find(({ crn }) => crn === parseInt(ctd$.children().eq(0).text()));
                        if (wlitem != null) {
                            wlitem.status = {
                                course: `${params.get("subj")}${params.get("crse")}`,
                                section: ctd$.children().eq(1).text().trim(),
                                webenabled: true,
                                avail: FULL_REGEX.test(ctd$.children().eq(6).text()) ? 0 : parseInt(ctd$.children().eq(6).text().replace(NUM_FILTER, "")),
                                cap: parseInt(ctd$.children().eq(7).text().replace(NUM_FILTER, "")),
                                waitlist: ctd$.children().eq(8).text().includes("N") ? false : (FULL_REGEX.test(ctd$.children().eq(8).text()) ? 0 : parseInt(ctd$.children().eq(8).text().replace(NUM_FILTER, ""))),
                                conflict: false,
                                updated: new Date().getTime()
                            };
                        }
                    }
                    if (ttb.classes.findIndex(u => u.crn === parseInt(ctd$.children().eq(0).text())) !== -1) {
                        td1$.css("color", "red").text(`Course registered`);
                    } else {
                        if (FULL_REGEX.test(ctd$.children().eq(6).text())) {
                            if (FULL_REGEX.test(ctd$.children().eq(8).text())) {
                                td1$.css("color", "red").text("Section is full, waitlist full");

                            } else if (ctd$.children().eq(8).text().includes("N")) {
                                td1$.css("color", "red").text("Section is full, waitlist not available");
                            } else {
                                if ($(v).children().eq(11).text().trim() === "") {
                                    td1$.css("color", "darkorange").html("Section is full, waitlist available<br />Registrable");
                                    if ($(v).children().eq(0).text().trim() !== "" && inAddDropPeriod) {
                                        td2$.append($("<a></a>").text("Add to Wishlist").attr({ "data-crn": $(v).children().eq(0).text(), "href": "#" }).click(addCRNToWishlist));
                                    }
                                } else {
                                    let day = DAYS.indexOf($(v).children().eq(10).text());
                                    let ztime = $(v).children().eq(11).text().split(" - ").map(t => new moment(t, "HH:mm"));
                                    if (selected === current) {
                                        y: for (let i of ttb.classes) {
                                            for (let j of i.times) {
                                                if (j === null || j.time === null) continue;
                                                let times = j.time.map(t => new moment(t, "hh:mm a"));
                                                if (day === j.day && (ztime.some(m => m.isBetween(times[0], times[1], undefined, "[]")) || times.some(m => m.isBetween(ztime[0], ztime[1], undefined, "[]")))) {
                                                    flag = false;
                                                    crash = `${i.name[1].replace(" ", "")} ${i.name[2]}`;
                                                    break y;
                                                }
                                            }
                                        }
                                        if (wlitem != null) {
                                            wlitem.status.conflict = flag ? false : crash;
                                        }
                                    }
                                    if (flag) {
                                        td1$.css("color", "darkorange").html(`Section is full, waitlist available<br />${selected === current ? "No conflicts" : "Incompatible terms"}`);
                                        if ($(v).children().eq(0).text().trim() !== "" && inAddDropPeriod) {
                                            td2$.append($("<a></a>").text("Add to Wishlist").attr({ "data-crn": $(v).children().eq(0).text(), "data-waitlist": "true", "href": "#" }).click(addCRNToWishlist));
                                        }
                                        if ($(v).children().eq(11).text().trim() !== "" && selected === current) {
                                            td3$.append($("<a></a>").text("Preview").attr("href", "#").click(preview));
                                        }
                                    } else {
                                        td1$.css("color", "red").html(`Section is full, waitlist available<br />Conflicts with ${crash}`);
                                        if ($(v).children().eq(0).text().trim() !== "" && inAddDropPeriod) {
                                            td2$.append($("<a></a>").text("Add to Wishlist").attr({ "data-crn": $(v).children().eq(0).text(), "data-crash": crash, "data-waitlist": "true", "href": "#" }).click(addCRNToWishlist));
                                        }
                                        if ($(v).children().eq(11).text().trim() !== "" && selected === current) {
                                            td3$.append($("<a></a>").text("Preview").attr("href", "#").click(preview));
                                        }
                                    }
                                }
                            }
                        } else {
                            if ($(v).children().eq(11).text().trim() === "") {
                                td1$.css("color", "green").text("Registrable");
                                if ($(v).children().eq(0).text().trim() !== "" && inAddDropPeriod) {
                                    td2$.append($("<a></a>").text("Add to Wishlist").attr({ "data-crn": $(v).children().eq(0).text(), "href": "#" }).click(addCRNToWishlist));
                                }
                            } else {
                                let day = DAYS.indexOf($(v).children().eq(10).text());
                                let ztime = $(v).children().eq(11).text().split(" - ").map(t => new moment(t, "HH:mm"));
                                if (selected === current) {
                                    x: for (let i of ttb.classes) {
                                        for (let j of i.times) {
                                            if (j === null || j.time === null) continue;
                                            let times = j.time.map(t => new moment(t, "hh:mm a"));
                                            if (day === j.day && (ztime.some(m => m.isBetween(times[0], times[1], undefined, "[]")) || times.some(m => m.isBetween(ztime[0], ztime[1], undefined, "[]")))) {
                                                flag = false;
                                                crash = `${i.name[1].replace(" ", "")} ${i.name[2]}`;
                                                break x;
                                            }
                                        }
                                    }
                                    if (wlitem != null) {
                                        wlitem.status.conflict = flag ? false : crash;
                                    }
                                }
                                if (flag) {
                                    td1$.css("color", selected === current ? "green" : "darkorange").text(selected === current ? "No conflicts" : "Incompatible terms");
                                    if ($(v).children().eq(0).text().trim() !== "" && inAddDropPeriod) {
                                        td2$.append($("<a></a>").text("Add to Wishlist").attr({ "data-crn": $(v).children().eq(0).text(), "href": "#" }).click(addCRNToWishlist));
                                    }
                                    if ($(v).children().eq(11).text().trim() !== "" && selected === current) {
                                        td3$.append($("<a></a>").text("Preview").attr("href", "#").click(preview));
                                    }
                                } else {
                                    td1$.css("color", "red").text(`Conflicts with ${crash}`);
                                    if ($(v).children().eq(0).text().trim() !== "" && inAddDropPeriod) {
                                        td2$.append($("<a></a>").text("Add to Wishlist").attr({ "data-crn": $(v).children().eq(0).text(), "data-crash": crash, "href": "#" }).click(addCRNToWishlist));
                                    }
                                    if ($(v).children().eq(11).text().trim() !== "") {
                                        td3$.append($("<a></a>").text("Preview").attr("href", "#").click(preview));
                                    }
                                }
                            }
                        }
                    }
                } else if ($(v).children().length === 16) {
                    if ($(v).children().eq(6).text().trim() !== "") {
                        ctd$ = $(v);
                        wlitem = wishlist.find(({ crn }) => crn === parseInt(ctd$.children().eq(0).text()));
                        if (wlitem != null) {
                            wlitem.status = {
                                course: `${params.get("subj")}${params.get("crse")}`,
                                section: ctd$.children().eq(1).text().trim(),
                                webenabled: false,
                                avail: null,
                                cap: null,
                                waitlist: null,
                                conflict: false,
                                updated: new Date().getTime()
                            };
                        }
                    }
                    if ($(v).children().eq(8).text().includes("N")) {
                        td1$.css("color", "red").text("Section not web-enabled");
                    }
                }
                if (inAddDropPeriod)
                    $(v).append([td1$, td2$, td3$]);
                else
                    $(v).append([td1$, td3$]);
            }
        });
        chrome.storage.local.set({ wishlist: wishlist });
    });
} else if (location.href === "https://banweb.cityu.edu.hk/pls/PROD/bwskfreg.P_AltPin") {
    chrome.storage.local.get(["wishlist", "autofill"], ({ wishlist, autofill }) => {
        if (autofill) {
            wishlist.filter(({ status }) => status == null || (status.webenabled && (status.waitlist || status.avail > 0))).map(v => v.crn).forEach((v, i) => {
                if (i < 10)
                    document.querySelectorAll("input[name='CRN_IN'][id]")[i].value = v;
            });
        }
    });
} else if (location.href === "https://banweb.cityu.edu.hk/pls/PROD/twgkutil_cityu.P_Search") {
    let a$ = $("a:contains('Master Class Schedule')");
    a$.get(a$.length - 1).click();
}

chrome.runtime.onMessage.addListener((m, s, r) => {
    console.log("on message", m);
    if (m.action === "mcs") {
        document.querySelector("#keyword_in_id").value = "Master Class Schedule";
        document.forms.cityu_search.submit();
    }
});

function preview() {
    let params = new URLSearchParams(location.search);
    let td$ = $(this).parents("tr");
    let ctd$ = td$;
    while (ctd$.children().eq(0).text().trim() === "") {
        ctd$ = ctd$.prev();
    }
    chrome.runtime.sendMessage({
        type: "preview",
        data: {
            course: `${params.get("subj")}${params.get("crse")}`,
            crn: parseInt(ctd$.children().eq(0).text()),
            day: DAYS.indexOf(td$.children().eq(10).text()),
            time: td$.children().eq(11).text().split(" - "),
            section: ctd$.children().eq(1).text(),
            instruct: ctd$.children().eq(14).text(),
            loc: `${ctd$.children().eq(12).text()} ${ctd$.children().eq(13).text()}`
        }
    });
    return false;
}

function addCRNToWishlist() {
    let params = new URLSearchParams(location.search);
    chrome.storage.local.get(["ttb", "wishlist"], ({ ttb, wishlist }) => {
        let ctd$ = $(this).parents("tr");
        let data = {
            crn: $(this).data("crn"),
            status: {
                course: `${params.get("subj")}${params.get("crse")}`,
                section: ctd$.children().eq(1).text().trim(),
                webenabled: true,
                avail: FULL_REGEX.test(ctd$.children().eq(6).text()) ? 0 : parseInt(ctd$.children().eq(6).text().replace(NUM_FILTER, "")),
                cap: parseInt(ctd$.children().eq(7).text().replace(NUM_FILTER, "")),
                waitlist: ctd$.children().eq(8).text().includes("N") ? false : (FULL_REGEX.test(ctd$.children().eq(8).text()) ? 0 : parseInt(ctd$.children().eq(8).text().replace(NUM_FILTER, ""))),
                conflict: $(this).data("crash") ?? false,
                updated: new Date().getTime()
            }
        };
        console.log("add to wishlist triggered with data:", data);
        if (wishlist == null) {
            wishlist = [];
        }
        if (ttb.classes.findIndex(v => v.crn === $(this).data("crn")) !== -1) {
            let f$ = $("<div></div>").css({ "font-size": "18px", "padding": "8px", "position": "fixed", "top": "20px", "right": "20px", "background-color": "white", "border": "5px solid red", "color": "red" }).text("This course is been already registered").appendTo(document.body);
            setTimeout(() => f$.remove(), 5000);
            return false;
        }
        if (wishlist.findIndex(({ crn }) => crn === $(this).data("crn")) !== -1) {
            let f$ = $("<div></div>").css({ "font-size": "18px", "padding": "8px", "position": "fixed", "top": "20px", "right": "20px", "background-color": "white", "border": "5px solid red", "color": "red" }).text("CRN already exist in the wishlist").appendTo(document.body);
            setTimeout(() => f$.remove(), 5000);
            return false;
        }
        if ($(this).data("crash")) {
            if (!confirm("This course section is conflicted with a course that you have been registered.\n\nYou must change/drop the conflicted course if you want to add this course.\n\nProceed anyway?"))
                return false;
        }
        if ($(this).data("waitlist")) {
            if (!confirm("This course section is full and waitlist available.\n\nYou will be put into the waitlist if you try to register this course.\n\nYOU MAY NOT BE ABLE TO REGISTER THIS COURSE AT THE END.\n\nProceed anyway?"))
                return false;
        }
        wishlist.push(data);
        chrome.storage.local.set({ wishlist: wishlist }, () => {
            let f$ = $("<div></div>").css({ "font-size": "18px", "padding": "8px", "position": "fixed", "top": "20px", "right": "20px", "background-color": "white", "border": "5px solid black" }).text("CRN added to wishlist").appendTo(document.body);
            setTimeout(() => f$.remove(), 5000);
        });
    });
    return false;
}